runtime/job-workers/client-functions.cpp

```diff
@@ -86,65 +105,127 @@ Optional<int64_t> f$kphp_job_worker_start(const class_instance<C$KphpJobWorkerRe
+  auto *memory_request = make_job_request_message<job_workers::JobSharedMessage>(request);
+  if (memory_request == nullptr) {
     return false;
   }
 
-  const auto now = std::chrono::system_clock::now();
-  memory_request->job_start_time = std::chrono::duration<double>{now.time_since_epoch()}.count();
-  memory_request->job_timeout = timeout;
+  int job_resumable_id = send_job_request_message(memory_request, timeout);
 
-  int job_id = 0;
-  {
-    dl::CriticalSectionSmartGuard critical_section;
-    job_id = client.send_job(memory_request);
-    if (job_id > 0) {
-      memory_manager.detach_shared_message_from_this_proc(memory_request);
+  if (job_resumable_id < 0) {
+    return false;
+  }
+  
+  return job_resumable_id;
+}
+
+array<Optional<int64_t>> f$kphp_job_worker_start_multi(const array<class_instance<C$KphpJobWorkerRequest>> &requests, double timeout) noexcept {
+  if (!job_workers_api_allowed()) {
+    return {};
+  }
+  timeout = normalize_job_timeout(timeout);
+
+  class_instance<C$KphpJobWorkerSharedMemoryPiece> common_shared_memory_piece;
+  bool first = true;
+  for (const auto &it : requests) {
+    const auto &req = it.get_value();
+    
+    if (req.is_null()) {
+      php_warning("Can't send multiple jobs: requests[%s] is null", it.get_key().to_string().c_str());
+      return {};
+    }
+    
+    const auto &cur_shared_mem_piece = req.get()->get_shared_memory_piece();
+    if (first) {
+      common_shared_memory_piece = cur_shared_mem_piece; // increment ref count here to prevent unexpected destroy on temporary resetting before copying below
+      first = false;
     } else {
-      memory_manager.release_shared_message(memory_request);
-      critical_section.leave_critical_section();
-      php_warning("Can't send job: probably jobs queue is full");
-      return false;
+      if (cur_shared_mem_piece.get() != common_shared_memory_piece.get()) {
+        php_warning("Can't extract common shared memory piece in multiple jobs request: requests[%s] and requests[%s] have different shared memory pieces",
+                    requests.begin().get_key().to_string().c_str(), it.get_key().to_string().c_str());
+        return {};
+      }
     }
   }
 
-  int64_t job_resumable_id = register_forked_resumable(new job_resumable{job_id});
+  array<Optional<int64_t>> res{requests.size()};
+
+  job_workers::JobSharedMemoryPiece *common_job_request = nullptr;
+  if (!common_shared_memory_piece.is_null()) {
+    common_job_request = make_job_request_message<job_workers::JobSharedMemoryPiece>(common_shared_memory_piece);
+    /**
+     * common_job_request lifetime:
+     * 1. attaches to client process on creating in `acquire_shared_message()`
+     * 2. attaches to job worker process on job receiving in `job_parse_execute()`
+     * 3. detaches from job worker process on terminating in `release_shared_message()` recursively
+     * 4. detaches from client process at the end of this function in `release_shared_message()` recursively
```

Comment:
It seems to me that at the end of this function `release_shared_message()` detaches common_job_request without recursion

Reply:
yep, fixed comment